src/HotChocolate/Caching/src/Caching/CacheControlConstraintsOptimizer.cs

```diff
@@ -0,0 +1,134 @@
+using System.Runtime.CompilerServices;
+using HotChocolate.Execution.Processing;
+using HotChocolate.Types;
+using HotChocolate.Types.Introspection;
+using HotChocolate.Utilities;
+
+namespace HotChocolate.Caching;
+
+/// <summary>
+/// Computes the cache control constraints for an operation during compilation.
+/// </summary>
+internal sealed class CacheControlConstraintsOptimizer : IOperationOptimizer
```

Comment:
I moved the `CacheControlConstraints` calculation into an operation optimizer. This is essentially a hook into the operation compiler and allows us to store stuff on the compiled operation.

In this case I am storing the `CacheControlConstraints` on the operation.